diff --git a/CMakeLists.txt b/CMakeLists.txt
index dc5830f..af2d55c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -55,6 +55,8 @@ if (DEFINED ZL_JUCE_FORMATS)
 else ()
     if (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
         set(FORMATS VST3 AU Standalone)
+    elseif (ANDROID)
+        set(FORMATS Standalone)
     else ()
         set(FORMATS VST3 LV2 Standalone)
     endif ()
@@ -227,3 +229,57 @@ target_link_libraries("${PROJECT_NAME}" PRIVATE SharedCode)
 
 # Pass some config to GA (like our PRODUCT_NAME)
 include(GitHubENV)
+
+
+# begin JUCEAAP specific -->
+
+include_directories(
+        "${AAP_DIR}/include"
+        "${JUCE_DIR}/modules"
+)
+
+juce_add_modules(${AAP_JUCE_DIR}/aap-modules/aap_audio_processors)
+target_link_libraries(${PROJECT_NAME}
+        PRIVATE
+        aap_audio_processors
+)
+
+add_compile_definitions(
+        [[JUCEAAP_USE_UNIQUE_ID=1]]
+        [[JUCEAAP_HOSTED_PARAMETER=1]]
+        [[JUCEAAP_HAVE_AUDIO_PLAYHEAD_NEW_POSITION_INFO=1]]
+)
+
+if (ANDROID)
+    add_library("cpufeatures" STATIC "${ANDROID_NDK}/sources/android/cpufeatures/cpu-features.c")
+    set_source_files_properties("${ANDROID_NDK}/sources/android/cpufeatures/cpu-features.c" PROPERTIES COMPILE_FLAGS "-Wno-sign-conversion -Wno-gnu-statement-expression")
+
+    # dependencies
+    find_library(log "log")
+    find_library(android "android")
+    find_library(glesv2 "GLESv2")
+    find_library(egl "EGL")
+    find_package(androidaudioplugin CONFIG REQUIRED)
+    set(cpufeatures_lib "cpufeatures")
+    set(oboe_lib "oboe")
+    set(CMAKE_CXX_FLAGS_RELEASE "-O3")
+
+    add_compile_definitions(
+            [[JUCE_ANDROID=1]]
+            [[JUCE_DONT_AUTO_OPEN_MIDI_DEVICES_ON_MOBILE=1]]
+            [[JUCE_PUSH_NOTIFICATIONS_ACTIVITY="com/rmsl/juce/JuceActivity"]]
+            [[JUCE_PUSH_NOTIFICATIONS=1]]
+            [[JUCE_USE_ANDROID_OBOE_STABILIZED_CALLBACK=1]]
+    )
+
+    target_link_libraries(${PROJECT_NAME}
+            PRIVATE
+            ${log}
+            ${android}
+            ${glesv2}
+            ${egl}
+            ${cpufeatures_lib}
+            androidaudioplugin::androidaudioplugin
+    )
+endif (ANDROID)
+# <-- end JUCEAAP specific
diff --git a/source/panel/ui_setting_panel/colour_setting_panel.cpp b/source/panel/ui_setting_panel/colour_setting_panel.cpp
index 47c1ddf..fc01e7b 100644
--- a/source/panel/ui_setting_panel/colour_setting_panel.cpp
+++ b/source/panel/ui_setting_panel/colour_setting_panel.cpp
@@ -10,6 +10,17 @@
 #include "colour_setting_panel.hpp"
 
 namespace zlpanel {
+    const juce::File& ColourSettingPanel::getSettingDirectory() {
+        static juce::File dir{};
+        if (dir.getFullPathName().isEmpty())
+            dir = juce::File::getSpecialLocation(juce::File::userApplicationDataDirectory)
+                    .getChildFile("Audio")
+                    .getChildFile("Presets")
+                    .getChildFile(JucePlugin_Manufacturer)
+                    .getChildFile("Shared Settings");
+        return dir;
+    }
+
     static juce::Colour getIntColour(const int r, const int g, const int b, float alpha) {
         return {
             static_cast<juce::uint8>(r),
@@ -23,8 +34,8 @@ namespace zlpanel {
         : pRef(p), base_(base), name_laf_(base),
           c_map1_selector_(base), c_map2_selector_(base) {
         juce::ignoreUnused(pRef);
-        if (!kSettingDirectory.isDirectory()) {
-            const auto result = kSettingDirectory.createDirectory();
+        if (!getSettingDirectory().isDirectory()) {
+            const auto result = getSettingDirectory().createDirectory();
             juce::ignoreUnused(result);
         }
         name_laf_.setFontScale(zlgui::kFontHuge);
@@ -140,7 +151,7 @@ namespace zlpanel {
     void ColourSettingPanel::mouseDown(const juce::MouseEvent &event) {
         if (event.originalComponent == &import_label_) {
             chooser_ = std::make_unique<juce::FileChooser>(
-                "Load the colour settings...", kSettingDirectory, "*.xml",
+                "Load the colour settings...", getSettingDirectory(), "*.xml",
                 true, false, nullptr);
             constexpr auto setting_open_flags = juce::FileBrowserComponent::openMode |
                                                 juce::FileBrowserComponent::canSelectFiles;
@@ -164,7 +175,7 @@ namespace zlpanel {
             });
         } else if (event.originalComponent == &export_label_) {
             chooser_ = std::make_unique<juce::FileChooser>(
-                "Save the colour settings...", kSettingDirectory.getChildFile("colour.xml"), "*.xml",
+                "Save the colour settings...", getSettingDirectory().getChildFile("colour.xml"), "*.xml",
                 true, false, nullptr);
             constexpr auto setting_save_flags = juce::FileBrowserComponent::saveMode |
                                                 juce::FileBrowserComponent::warnAboutOverwriting;
diff --git a/source/panel/ui_setting_panel/colour_setting_panel.hpp b/source/panel/ui_setting_panel/colour_setting_panel.hpp
index 0042032..98083e9 100644
--- a/source/panel/ui_setting_panel/colour_setting_panel.hpp
+++ b/source/panel/ui_setting_panel/colour_setting_panel.hpp
@@ -49,11 +49,6 @@ namespace zlpanel {
         juce::Label import_label_, export_label_;
 
         std::unique_ptr<juce::FileChooser> chooser_;
-        inline auto static const kSettingDirectory =
-                juce::File::getSpecialLocation(juce::File::userApplicationDataDirectory)
-                .getChildFile("Audio")
-                .getChildFile("Presets")
-                .getChildFile(JucePlugin_Manufacturer)
-                .getChildFile("Shared Settings");
+        static const juce::File& getSettingDirectory();
     };
 } // zlpanel
diff --git a/source/panel/ui_setting_panel/control_setting_panel.cpp b/source/panel/ui_setting_panel/control_setting_panel.cpp
index 1859191..3614f2e 100644
--- a/source/panel/ui_setting_panel/control_setting_panel.cpp
+++ b/source/panel/ui_setting_panel/control_setting_panel.cpp
@@ -10,6 +10,17 @@
 #include "control_setting_panel.hpp"
 
 namespace zlpanel {
+    const juce::File& ControlSettingPanel::getSettingDirectory() {
+        static juce::File dir{};
+        if (dir.getFullPathName().isEmpty())
+            dir = juce::File::getSpecialLocation(juce::File::userApplicationDataDirectory)
+                    .getChildFile("Audio")
+                    .getChildFile("Presets")
+                    .getChildFile(JucePlugin_Manufacturer)
+                    .getChildFile("Shared Settings");
+        return dir;
+    }
+
     ControlSettingPanel::ControlSettingPanel(PluginProcessor &p, zlgui::UIBase &base)
         : p_ref_(p),
           base_(base), name_laf_(base),
@@ -165,7 +176,7 @@ namespace zlpanel {
 
     void ControlSettingPanel::importControls() {
         chooser_ = std::make_unique<juce::FileChooser>(
-            "Load the control settings...", kSettingDirectory, "*.xml",
+            "Load the control settings...", getSettingDirectory(), "*.xml",
             true, false, nullptr);
         constexpr auto setting_open_flags = juce::FileBrowserComponent::openMode |
                                             juce::FileBrowserComponent::canSelectFiles;
@@ -213,7 +224,7 @@ namespace zlpanel {
 
     void ControlSettingPanel::exportControls() {
         chooser_ = std::make_unique<juce::FileChooser>(
-            "Save the control settings...", kSettingDirectory.getChildFile("control.xml"), "*.xml",
+            "Save the control settings...", getSettingDirectory().getChildFile("control.xml"), "*.xml",
             true, false, nullptr);
         constexpr auto setting_save_flags = juce::FileBrowserComponent::saveMode |
                                             juce::FileBrowserComponent::warnAboutOverwriting;
diff --git a/source/panel/ui_setting_panel/control_setting_panel.hpp b/source/panel/ui_setting_panel/control_setting_panel.hpp
index b7b8163..5c96c6f 100644
--- a/source/panel/ui_setting_panel/control_setting_panel.hpp
+++ b/source/panel/ui_setting_panel/control_setting_panel.hpp
@@ -51,12 +51,7 @@ namespace zlpanel {
 
         juce::Label import_label_, export_label_;
         std::unique_ptr<juce::FileChooser> chooser_;
-        inline auto static const kSettingDirectory =
-                juce::File::getSpecialLocation(juce::File::userApplicationDataDirectory)
-                .getChildFile("Audio")
-                .getChildFile("Presets")
-                .getChildFile(JucePlugin_Manufacturer)
-                .getChildFile("Shared Settings");
+        static const juce::File& getSettingDirectory();
 
         void importControls();
 
